components/brave_wallet/browser/meld_integration_response_parser.cc

```diff
@@ -0,0 +1,408 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/meld_integration_response_parser.h"
+
+#include <algorithm>
+#include <optional>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include "base/ranges/algorithm.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/types/expected.h"
+#include "brave/components/brave_wallet/browser/meld_integration_responses.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom-forward.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "tools/json_schema_compiler/util.h"
+
+namespace {
+
+std::optional<brave_wallet::mojom::MeldLogoImagesPtr> ParseMeldLogos(
+    const std::optional<brave_wallet::meld_integration_responses::Logos>&
+        logos) {
+  if (!logos) {
+    return std::nullopt;
+  }
+
+  auto logo_images = brave_wallet::mojom::MeldLogoImages::New(
+      logos->dark, logos->dark_short, logos->light, logos->light_short);
+
+  return std::move(logo_images);
+}
+
+std::optional<std::vector<brave_wallet::mojom::MeldRegionPtr>> ParseMeldRegions(
+    const std::optional<
+        std::vector<brave_wallet::meld_integration_responses::Region>>&
+        regions) {
+  if (!regions) {
+    return std::nullopt;
+  }
+
+  std::vector<brave_wallet::mojom::MeldRegionPtr> result;
+  for (const auto& region_value : *regions) {
+    auto reg = brave_wallet::mojom::MeldRegion::New(region_value.region_code,
+                                                    region_value.name);
+    result.emplace_back(std::move(reg));
+  }
+
+  if (result.empty()) {
+    return std::nullopt;
+  }
+
+  return result;
+}
+
+std::optional<base::flat_map<std::string, std::string>>
+ParseOptionalMapOfStrings(const std::optional<base::Value>& val) {
+  if (!val || !val->is_dict()) {
+    return std::nullopt;
+  }
+
+  base::flat_map<std::string, std::string> result;
+  for (const auto [key, value] : val->GetDict()) {
+    result.insert_or_assign(key, value.GetString());
+  }
+  if (result.empty()) {
```

Comment:
Inserting into flat_map in a loop is a O(N^2). What is expected maximum size of such collection?

Reply:
according to the current documentation it is 3 elements